### DISCUSSION TITLE
Get RevisedSamplingRate from server

### DISCUSSION THREAD

**Discussion by DasCapschen** [FreeOpcUa/opcua-asyncio#1523] [Q&A] [answered]
Hello,

I currently have the problem that some data from the OPC UA server I'm subscribed to does not arrive in my application, but my customer does receive the data.
It's a machine error code node that quickly changes its value, let me give an example:

1. node value = 0 -> machine running
2. node value = 5 -> real stop error of the machine
3. node value = 7 -> non-helpful follow-up error

In my application, I only see node values 0 and 7, but 5 never arrives. 
The problem is, my customer, who is using a sampling rate of no faster than 100ms, does receive the value.

I'm using `client.create_subscription()` with `period = 0` for the publish interval, which means as fast as the server can handle, and I already saw that I can check the RevisedPublishRate in the returned subscription.  
This doesn't help me, because the Publish and Sampling interval are completely independent.

I'm then using `subscription.subscribe_data_change()` with `sampling_interval = 0`, which should be the fastest sampling interval the server supports.  
Unfortunately, I was unable to find out how I can check the RevisedSamplingInterval of this data change subscription. 
Could you let me know if it is possible to find out which sampling rate the server actually uses? 

Thank you.

**Comment by AndreasHeine** [comment]
even with period ("PublishingInterval") 1s you would recv the change as long the queue is not full!
the period in create subscripion just tells you how often the queued changes will be transported...

https://github.com/AndreasHeine/opcua-tutorial/blob/main/client/subscribe.py#L43-L60

edit:

the revised params will be available in the subscription-class-instance itself:
https://github.com/FreeOpcUa/opcua-asyncio/blob/bec29074dac1482a6e759fa1d60b0dfba1d61774/asyncua/client/client.py#L746

**Comment by DasCapschen** [comment]
My Problem is not with the PublishInterval (period), but with the sampling rate.  
I suspect the SamplingRate used is larger than the time interval between changes. 

example:
```
v=0  v=0      v=5    v=7
|    |        |      |
t=0  t=5      t=7    t=10
     |               |
    sampling        sampling
```

In this case I would never receive value = 5, because it changed faster than the OPCUA Server samples data changes.      
So I'm not looking for the revised publish interval (i already found that), but the revised sampling interval (from `subscription.subscribe_data_change()`)

**Reply by AndreasHeine** [reply]
what is the MinimumSamplingInterval (Attribute) of that Variable?

**Reply by DasCapschen** [reply]
You mean the MinimumSamplingInterval of the Node I am monitoring?  
I checked with `node.read_attribute()` and got a Value of `5.0` (I assume this is Milliseconds?), was this what you were asking?

**Reply by AndreasHeine** [reply]
https://reference.opcfoundation.org/Core/Part4/v105/docs/5.12.1.2

The optional [MinimumSamplingInterval Attribute](https://reference.opcfoundation.org/search/18?t=MinimumSamplingInterval%20Attribute) defined in [OPC 10000-3](https://reference.opcfoundation.org/Core/Part4/v105/docs/?r=UAPart3) identifies the minimum sampling interval supported for a [Variable](https://reference.opcfoundation.org/search/18?t=Variable).

https://reference.opcfoundation.org/Core/Part3/v105/docs/5.6

The MinimumSamplingInterval Attribute indicates how “current” the Value of the Variable will be kept. It specifies (in milliseconds) how fast the Server can reasonably sample the value for changes (see OPC 10000-4 for a detailed description of sampling interval).

A MinimumSamplingInterval of 0 indicates that the Server is to monitor the item continuously. A MinimumSamplingInterval of −1means indeterminate.


> So your revised value should be 5.0!?
> there should be a logger entry like:
> _logger.info(f"Result from subscription update: {results}")
> however if the change is faster then the MinimumSamplingInterval you will not detect it...

**Reply by DasCapschen** [reply]
The revised value does not need to be 5, because as per https://reference.opcfoundation.org/Core/Part4/v105/docs/5.12.1.2:
> If the Server specifies a value for the MinimumSamplingInterval Attribute it shall always return a revisedSamplingInterval that is equal or higher than the MinimumSamplingInterval if the Client subscribes to the Value Attribute.

The server can return a value larger than MinimumSamplingInterval.

I will check if I can see any logs and let you know what the logs say.

**Reply by AndreasHeine** [reply]
thats true but if the server returns a higher value it has usually issues with performance...

**Reply by DasCapschen** [reply]
The `result from subscription update` happens right after subscribe, and only contains the revised subscription parameters, like the `RevisedPublishInterval`, but not the revised **sampling** interval.

**Reply by DasCapschen** [reply]
It would be nice to know if I can see somewhere **if** it returns a higher value. If performance problems are the cause, that's OK, but I would like to at least know that :)

**Reply by DasCapschen** [reply]
Happy new year :)
Just wanted to ask again, perhaps you didn't see the previous comment ; is there a way to see *if* a higher value is returned for the revised sampling interval?

Thank you

**Reply by AndreasHeine** [reply]
hey, did you use the UaClient or the Client Class?

**Reply by AndreasHeine** [reply]
With the UaClient calls its like following

For the subscription:
https://github.com/FreeOpcUa/opcua-asyncio/blob/b9e2af4e8d079b3a0c4eb6a6e49f861e8e309fb7/asyncua/client/ua_client.py#L500

returns:
https://github.com/FreeOpcUa/opcua-asyncio/blob/b9e2af4e8d079b3a0c4eb6a6e49f861e8e309fb7/asyncua/ua/uaprotocol_auto.py#L8646

contains:
https://github.com/FreeOpcUa/opcua-asyncio/blob/b9e2af4e8d079b3a0c4eb6a6e49f861e8e309fb7/asyncua/ua/uaprotocol_auto.py#L8627

For the MonitoredItem:
https://github.com/FreeOpcUa/opcua-asyncio/blob/b9e2af4e8d079b3a0c4eb6a6e49f861e8e309fb7/asyncua/client/ua_client.py#L643

returns:
https://github.com/FreeOpcUa/opcua-asyncio/blob/b9e2af4e8d079b3a0c4eb6a6e49f861e8e309fb7/asyncua/ua/uaprotocol_auto.py#L8153

contains:
https://github.com/FreeOpcUa/opcua-asyncio/blob/b9e2af4e8d079b3a0c4eb6a6e49f861e8e309fb7/asyncua/ua/uaprotocol_auto.py#L8067

With the Client calls its like following

For the subscription:
https://github.com/FreeOpcUa/opcua-asyncio/blob/b9e2af4e8d079b3a0c4eb6a6e49f861e8e309fb7/asyncua/client/client.py#L721
returns:
https://github.com/FreeOpcUa/opcua-asyncio/blob/b9e2af4e8d079b3a0c4eb6a6e49f861e8e309fb7/asyncua/common/subscription.py#L134

For the MonitoredItem:
https://github.com/FreeOpcUa/opcua-asyncio/blob/b9e2af4e8d079b3a0c4eb6a6e49f861e8e309fb7/asyncua/common/subscription.py#L259
returns the info

**Reply by DasCapschen** [reply]
I'm using the Client class. `subscribe_data_change` only returns ints as handles to the monitored items, not actually the revised parameters returned from the server, or do I misunderstand that?

But I could try to use `self.uaclient` of the Client class to create the monitored items myself and then check the result.

**Reply by AndreasHeine** [reply, marked as answer]
my fault your right...

so you need to rewrite the subscrition.subscribe_data_change part to client.uaclient.create_monitored_items()

https://github.com/FreeOpcUa/opcua-asyncio/blob/b9e2af4e8d079b3a0c4eb6a6e49f861e8e309fb7/asyncua/client/ua_client.py#L643

similar to:
https://github.com/FreeOpcUa/opcua-asyncio/blob/b9e2af4e8d079b3a0c4eb6a6e49f861e8e309fb7/asyncua/common/subscription.py#L450

**Reply by AndreasHeine** [reply]
the thing is the Client-Class was ment to be highlevel for ease of use but it lacks some details which in most situations aren't important...

**Reply by DasCapschen** [reply]
Thanks! I will give that a try :)
